client/rest/src/main/java/org/elasticsearch/client/RequestOptions.java

```diff
@@ -0,0 +1,175 @@
+ * can be manipulated without changing Elasticsearch's behavior.
+ */
+public final class RequestOptions {
+    public static final RequestOptions DEFAULT = new Builder(
+        Collections.<Header>emptyList(), HeapBufferedResponseConsumerFactory.DEFAULT).build();
+
+    private final List<Header> headers;
+    private final HttpAsyncResponseConsumerFactory httpAsyncResponseConsumerFactory;
+
+    private RequestOptions(Builder builder) {
+        this.headers = Collections.unmodifiableList(new ArrayList<>(builder.headers));
+        this.httpAsyncResponseConsumerFactory = builder.httpAsyncResponseConsumerFactory;
+    }
+
+    public Builder toBuilder() {
+        Builder builder = new Builder(headers, httpAsyncResponseConsumerFactory);
+        return builder;
+    }
+
+    /**
+     * Headers to attach to the request.
+     */
+    public List<Header> getHeaders() {
+        return headers;
+    }
+
+    /**
+     * The {@link HttpAsyncResponseConsumerFactory} used to create one
+     * {@link HttpAsyncResponseConsumer} callback per retry. Controls how the
+     * response body gets streamed from a non-blocking HTTP connection on the
+     * client side.
+     */
+    public HttpAsyncResponseConsumerFactory getHttpAsyncResponseConsumerFactory() {
+        return httpAsyncResponseConsumerFactory;
+    }
+
+    @Override
+    public String toString() {
+        StringBuilder b = new StringBuilder();
+        b.append("RequestOptions{");
+        if (headers.size() > 0) {
+            b.append(", headers=");
+            for (int h = 0; h < headers.size(); h++) {
+                if (h != 0) {
+                    b.append(',');
+                }
+                b.append(headers.get(h).toString());
+            }
+        }
+        if (httpAsyncResponseConsumerFactory != HttpAsyncResponseConsumerFactory.DEFAULT) {
+            b.append(", consumerFactory=").append(httpAsyncResponseConsumerFactory);
+        }
+        return b.append('}').toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || (obj.getClass() != getClass())) {
+            return false;
+        }
+        if (obj == this) {
+            return true;
+        }
+
+        RequestOptions other = (RequestOptions) obj;
+        return headers.equals(other.headers)
+                && httpAsyncResponseConsumerFactory.equals(other.httpAsyncResponseConsumerFactory);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(headers, httpAsyncResponseConsumerFactory);
+    }
+
+    public static class Builder {
+        private final List<Header> headers;
+        private HttpAsyncResponseConsumerFactory httpAsyncResponseConsumerFactory;
+
+        private Builder(List<Header> headers, HttpAsyncResponseConsumerFactory httpAsyncResponseConsumerFactory) {
+            this.headers = new ArrayList<>(headers);
+            this.httpAsyncResponseConsumerFactory = httpAsyncResponseConsumerFactory;
```

Comment:
why make this also settable if it's provided at construction time? Also should we check that it's not null?

Reply:
The idea is that you can make a `RequestOptions` that you use for all requests and then tweak it for certain requests. So the builder has to start with whatever was in the `RequestOptions` that you call `toBuilder` on but you can modify it from there.

Reply:
gotcha, shall we also have an empty constructor for cases where users want to start from scratch?

Reply:
I was doing that with `RequestOptions.DEFAULTS.toBuilder()`. I kind of like how explicit it is about starting with the defaults.

Reply:
I see, I wonder if having `new RequestOptions.Builder()` or `RequestOptions.builder()` would make this simpler from a users perspective.

Reply:
I think `RequestOptions.DEFAULTS.toBuilder()` is more obvious when you read it than `RequestOptions.builder()`.